File: ObjectProgrammingPython/Zwierzeta/czlowiek.py
from Obsluga.para import Para
from Zwierzeta.zwierze import Zwierze
from organizm import Organizm


class Czlowiek(Zwierze):
    def __init__(self, Swiat, x, y, czyZyje):
        super(Czlowiek, self).__init__(Swiat, x, y, czyZyje)
        self.sila = 5
        self.inicjatywa = 4
        self.kolor = "yellow"
        self.oznaczenie = "C"
        self.czyMozeRozmnazac = 0
        self.wiek = 0

    def akcja(self):
        self.przygotujOrganizm()
        self.ustawRuch()
        status = ""
        kom = self.swiat.getKomentarz()
        if self.swiat.getCzyUmiejetnosc():
            self.swiat.setKtoraTuraUmiejetnosci(self.swiat.getKtoraTuraUmiejetnosci() + 1)
            if self.swiat.getKtoraTuraUmiejetnosci() < 5:
                self.wywolajUmiejetnosc()
                status = self.Wypisz() + " WYWOLAL UMIEJETNOSC CALOPALENIE, JEST TO " + \
                         str(self.swiat.getKtoraTuraUmiejetnosci()) + " TURA UMIEJETNOSCI"
        else:
            status = self.Wypisz() + " UMIEJETNOSC SKONCZYLA SIE"
            self.swiat.setCzyUmiejetnosc(False)
            self.swiat.setKtoraTuraUmiejetnosci(0)
        if status != "":
            kom.append(status)

    def ustawRuch(self):
        self.przygotujOrganizm()
        ruch = self.swiat.getOstatniKlawisz()
        status = ""
        kom = self.swiat.getKomentarz()
        swiatOrganizmow = self.swiat.getSwiatOrganizmow()
        if ruch == 1:
            if self.polozenieX + 1 < self.swiat.getSizeX():
                if swiatOrganizmow[self.polozenieX + 1][self.polozenieY] is not None:
                    swiatOrganizmow[self.polozenieX + 1][self.polozenieY].kolizja(self)
                else:
                    swiatOrganizmow[self.polozenieX][self.polozenieY] = None
                    status = self.Wypisz() + " PORUSZYL SIE NA POZYCJE " + str(self.polozenieX + 1) + ' ' + str(
                        self.polozenieY)
                    self.polozenieX += 1
                    swiatOrganizmow[self.polozenieX][self.polozenieY] = self
        elif ruch == 0:
            if self.polozenieX > 0:
                if swiatOrganizmow[self.polozenieX - 1][self.polozenieY] is not None:
                    swiatOrganizmow[self.polozenieX - 1][self.polozenieY].kolizja(self)
                else:
                    swiatOrganizmow[self.polozenieX][self.polozenieY] = None
                    status = self.Wypisz() + " PORUSZYL SIE NA POZYCJE " + str(self.polozenieX - 1) + ' ' + str(
                        self.polozenieY)
                    self.polozenieX -= 1
                    swiatOrganizmow[self.polozenieX][self.polozenieY] = self
        elif ruch == 3:
            if self.polozenieY + 1 < self.swiat.getSizeY():
                if swiatOrganizmow[self.polozenieX][self.polozenieY + 1] is not None:
                    swiatOrganizmow[self.polozenieX][self.polozenieY + 1].kolizja(self)
                else:
                    swiatOrganizmow[self.polozenieX][self.polozenieY] = None
                    status = self.Wypisz() + " PORUSZYL SIE NA POZYCJE " + str(self.polozenieX) + ' ' + str(
                        self.polozenieY + 1)
                    self.polozenieY += 1
                    swiatOrganizmow[self.polozenieX][self.polozenieY] = self
        elif ruch == 2:
            if self.polozenieY > 0:
                if swiatOrganizmow[self.polozenieX][self.polozenieY - 1] is not None:
                    swiatOrganizmow[self.polozenieX][self.polozenieY - 1].kolizja(self)
                else:
                    swiatOrganizmow[self.polozenieX][self.polozenieY] = None
                    status = self.Wypisz() + " PORUSZYL SIE NA POZYCJE " + str(self.polozenieX) + ' ' + str(
                        self.polozenieY - 1)
                    self.polozenieY -= 1
                    swiatOrganizmow[self.polozenieX][self.polozenieY] = self
        elif ruch == 4:
            self.swiat.setCzyUmiejetnosc(True)
        if status != "":
            kom.append(status)

    def wywolajUmiejetnosc(self):
        swiatOrganizmow = self.swiat.getSwiatOrganizmow()
        pelnePola = self.znajdzPelne(self.polozenieX, self.polozenieY)
        status = ""
        kom = self.swiat.getKomentarz()
        if len(pelnePola) > 0:
            for wsp in pelnePola:
                status = self.Wypisz() + " ZABIL ZA POMOCA UMIEJETNOSCI " + swiatOrganizmow[wsp.getX()][
                    wsp.getY()].Wypisz()
                swiatOrganizmow[wsp.getX()][wsp.getY()].setCzyZyje(False)
                swiatOrganizmow[wsp.getX()][wsp.getY()] = None
                kom.append(status)

    def znajdzPelne(self, x, y):
        swiatOrganizmow = self.swiat.getSwiatOrganizmow()
        sizeX = self.swiat.getSizeX()
        sizeY = self.swiat.getSizeY()
        listaPustych = []
        if x - 1 >= 0 and y - 1 >= 0 and swiatOrganizmow[x - 1][y - 1] is not None:
            listaPustych.append(Para(x - 1, y - 1))
        if x - 1 >= 0 and swiatOrganizmow[x - 1][y] is not None:
            listaPustych.append(Para(x - 1, y))
        if x - 1 >= 0 and y + 1 < sizeY and swiatOrganizmow[x - 1][y + 1] is not None:
            listaPustych.append(Para(x - 1, y + 1))
        if y - 1 >= 0 and swiatOrganizmow[x][y - 1] is not None:
            listaPustych.append(Para(x, y - 1))
        if y + 1 < sizeY and swiatOrganizmow[x][y + 1] is not None:
            listaPustych.append(Para(x, y + 1))
        if x + 1 < sizeX and y - 1 >= 0 and swiatOrganizmow[x + 1][y - 1] is not None:
            listaPustych.append(Para(x + 1, y - 1))
        if x + 1 < sizeX and swiatOrganizmow[x + 1][y] is not None:
            listaPustych.append(Para(x + 1, y))
        if x + 1 < sizeX and y + 1 < sizeY and swiatOrganizmow[x + 1][y + 1] is not None:
            listaPustych.append(Para(x + 1, y + 1))
        return listaPustych

    def Wypisz(self):
        return "Czlowiek" + self.koncowka()

File: ObjectProgrammingJava/projekt po - 2/src/projekt/zwierzeta/Czlowiek.java
package projekt.zwierzeta;

import projekt.Organizm;
import projekt.Wypisanie;
import projekt.obsluga.Para;
import projekt.swiat.Swiat;

import java.awt.*;
import java.util.Vector;

public class Czlowiek extends Zwierze implements Wypisanie {
    private static final int UP = 0;
    private static final int LEFT = 1;
    private static final int DOWN = 2;
    private static final int RIGHT = 3;
    private static final int SPACE = 4;

    public Czlowiek(Swiat swiat, int x, int y, boolean czyZyje){
        polozenieX = x;
        polozenieY = y;
        sila = 5;
        inicjatywa = 4;
        kolor = Color.yellow;
        oznaczenie = 'C';
        this.swiat = swiat;
        this.czyZyje = czyZyje;
        this.czyMozeRozmnazac = 0;
        wiek = 0;
    }

    @Override
    public void akcja(){
        wiek++;
        ustawRuch();
        String status = "";
        Vector<String> kom = swiat.getKomentarz();
        if (swiat.isCzyUmiejetnosc()) {
            swiat.setKtoraTuraUmiejetnosci(swiat.getKtoraTuraUmiejetnosci() + 1);
            if (swiat.getKtoraTuraUmiejetnosci() < 5) {
                wywolajUmiejetnosc();
                status = Wypisz() + " WYWOLAL UMIEJETNOSC CALOPALENIE, JEST TO " + String.valueOf(swiat.getKtoraTuraUmiejetnosci())+ " TURA UMIEJETNOSCI";
            }
            else {
                status = Wypisz() + " UMIEJETNOSC SKONCZYLA SIE";
                swiat.setCzyUmiejetnosc(false);
                swiat.setKtoraTuraUmiejetnosci(0);
            }
        }
        if (!status.equals("")) {
            kom.add(status);
        }
    }

    @Override
    public String Wypisz() {
        return "Czlowiek" + getKoncowka();
    }

    private void ustawRuch(){
        Organizm[][] swiatOrganizmow = swiat.getSwiatOrganizmow();
        int klawisz = swiat.getOstatniKlawisz();
        String status = "";
        Vector<String> kom = swiat.getKomentarz();
        switch (klawisz)
        {
            case UP:
                if (polozenieX > 0) {
                    if(swiatOrganizmow[polozenieX-1][polozenieY] != null){
                        swiatOrganizmow[polozenieX-1][polozenieY].kolizja(this);
                    }else{
                        swiatOrganizmow[polozenieX][polozenieY] = null;
                        status = this.Wypisz() + " PORUSZYL SIE NA POZYCJE ";
                        polozenieX--;
                        status +=  String.valueOf(polozenieX)+ ":" + String.valueOf(polozenieY);
                        swiatOrganizmow[polozenieX][polozenieY] = this;
                    }
                }
                break;
            case RIGHT:
                if (polozenieY+1 < swiat.getSizeY()) {
                    if(swiatOrganizmow[polozenieX][polozenieY+1] != null){
                        swiatOrganizmow[polozenieX][polozenieY+1].kolizja(this);
                    }else {
                        swiatOrganizmow[polozenieX][polozenieY] = null;
                        status = this.Wypisz() + " PORUSZYL SIE NA POZYCJE ";
                        polozenieY++;
                        status += String.valueOf(polozenieX) + ":" + String.valueOf(polozenieY);
                        swiatOrganizmow[polozenieX][polozenieY] = this;
                    }
                }
                break;
            case DOWN:
                if (polozenieX+1 < swiat.getSizeX()) {
                    if(swiatOrganizmow[polozenieX+1][polozenieY] != null){
                        swiatOrganizmow[polozenieX+1][polozenieY].kolizja(this);
                    }else{
                        swiatOrganizmow[polozenieX][polozenieY] = null;
                        status = this.Wypisz() + " PORUSZYL SIE NA POZYCJE ";
                        polozenieX++;
                        status +=  String.valueOf(polozenieX)+ ":" + String.valueOf(polozenieY);
                        swiatOrganizmow[polozenieX][polozenieY] = this;
                    }
                }
                break;
            case LEFT:
                if (polozenieY > 0) {
                    if(swiatOrganizmow[polozenieX][polozenieY-1] != null){
                        swiatOrganizmow[polozenieX][polozenieY-1].kolizja(this);
                    }else{
                        swiatOrganizmow[polozenieX][polozenieY] = null;
                        status = this.Wypisz() + " PORUSZYL SIE NA POZYCJE ";
                        polozenieY--;
                        status +=  String.valueOf(polozenieX)+ ":" + String.valueOf(polozenieY);
                        swiatOrganizmow[polozenieX][polozenieY] = this;
                    }
                }
                break;
            case SPACE:
                swiat.setCzyUmiejetnosc(true);
                break;
            default:
                break;
        }
        if (!status.equals("")) {
            kom.add(status);
        }
    }

    private void wywolajUmiejetnosc(){
        Organizm[][] swiatOrganizmow = swiat.getSwiatOrganizmow();
        Vector<Para> pelnePola = znajdzPelne(polozenieX, polozenieY);
        String status;
        Vector<String>  kom = swiat.getKomentarz();
        if (pelnePola.size() > 0) {
            for (Para wsp : pelnePola) {
                status = Wypisz() + " ZABIL ZA POMOCA UMIEJETNOSCI " + swiatOrganizmow[wsp.getL()][wsp.getR()].Wypisz();
                swiatOrganizmow[wsp.getL()][wsp.getR()].setCzyZyje(false);
                swiatOrganizmow[wsp.getL()][wsp.getR()] = null;
                kom.add(status);
            }
        }
    }

    public Vector<Para> znajdzPelne(int x, int y) {
        Organizm[][] swiatOrganizmow = swiat.getSwiatOrganizmow();
        int sizeX = swiat.getSizeX();
        int sizeY = swiat.getSizeY();
        Vector<Para> vect = new Vector<>();
        if (x - 1 >= 0 && y - 1 > 0 && swiatOrganizmow[x - 1][y - 1] != null) {
            vect.add(new Para(x-1, y-1));
        }
        if (x - 1 >= 0 && swiatOrganizmow[x - 1][y] != null) {
            vect.add(new Para(x-1,y));
        }
        if (x - 1 >= 0 && y + 1 < sizeY && swiatOrganizmow[x - 1][y + 1] != null) {
            vect.add(new Para(x-1,y +1));
        }
        if (y - 1 >= 0 && swiatOrganizmow[x][y - 1] != null) {
            vect.add(new Para(x,y - 1));
        }
        if (y + 1 < sizeY && swiatOrganizmow[x][y + 1] != null) {
            vect.add(new Para(x,y +1));
        }
        if (x + 1 < sizeX && y - 1 >= 0 && swiatOrganizmow[x + 1][y - 1] != null) {
            vect.add(new Para(x+1,y - 1));
        }
        if (x + 1 < sizeX && swiatOrganizmow[x + 1][y] != null) {
            vect.add(new Para(x+1,y));
        }
        if (x + 1 < sizeX && y + 1 < sizeY && swiatOrganizmow[x + 1][y + 1] != null) {
            vect.add(new Para(x+1,y + 1));
        }
        return vect;
    }
}
